example/src/Main.purs

```diff
@@ -7,86 +7,98 @@ import Data.Void
-  w <- document globalWindow
-  b <- body w
-  appendChild b e
 
-data State = State String SlamDownState
+data Input a = DocumentChanged String a
 
-data Input
-  = DocumentChanged String
-  | FormFieldChanged SlamDownEvent
+type ChildSlots = Unit
 
-ui :: forall f eff. (Alternative f) => BrowserFeatures -> Component f Input Input
-ui browserFeatures = render <$> stateful (State "" (initSlamDownState $ SlamDown mempty)) update
+instance functorInput :: Functor Input where
+  map f (DocumentChanged d a) = DocumentChanged d $ f a
+
+parent :: forall g p. (Functor g) => H.ParentComponentP State SlamDownState Input SlamDownInput g ChildSlots p
+parent = H.component' render eval peek
   where
-  render :: State -> H.HTML (f Input)
-  render (State md form) =
-    H.div [ A.class_ (A.className "container") ]
-          [ H.h2_ [ H.text "Markdown" ]
-          , H.div_ [ H.textarea [ A.class_ (A.className "form-control")
-                                , A.value md
-                                , A.onInput (A.input DocumentChanged)
-                                ] [] ]
-          , H.h2_ [ H.text "HTML Output" ]
-          , H.div [ A.class_ (A.className "well") ] (output md form)
-          , H.h2_ [ H.text "Form State" ]
-          , H.pre_ [ H.code_ [ H.text (show form) ] ]
-          ]
-
-  output :: String -> SlamDownState -> Array (H.HTML (f Input))
-  output md form = ((FormFieldChanged <$>) <$>) <$> renderHalogen browserFeatures "slamdown-example" form (parseMd md)
-
-  update :: State -> Input -> State
-  update (State _ (SlamDownState form)) (DocumentChanged md) =
-    let slamdown = parseMd md
-    in case initSlamDownState slamdown of
-      SlamDownState emptyForm ->
-        let currentKeys = SM.keys emptyForm
-            oldKeys = SM.keys form \\ currentKeys
-            newForm = foldl (\f k -> SM.delete k f) (form `SM.union` emptyForm) oldKeys
-        in State md (SlamDownState newForm)
-  update (State s form) (FormFieldChanged e) = State s (applySlamDownEvent form e)
-
-main = onLoad do
+    render :: H.Render State Input ChildSlots
+    render state = do
+      H.div
+        [ P.class_ $ H.className "container" ]
+        [ H.h2_ [ H.text "Markdown" ]
+        , H.div_
+            [ H.textarea
+                [ P.class_ $ H.className "form-control"
+                , P.value state.markdown
+                , E.onValueInput $ E.input DocumentChanged
+                ]
+            ]
+        , H.h2_ [ H.text "HTML Output" ]
+        , H.div [ P.class_ (H.className "well") ] [ H.slot unit ]
+        , H.h2_ [ H.text "Form State" ]
+        , H.pre_ [ H.code_ [ H.text (show state.formState) ] ]
+        ]
+
+    eval :: H.EvalP Input State SlamDownState Input SlamDownInput g ChildSlots p
+    eval (DocumentChanged text next) = do
+      H.query unit <<< H.action <<< SetDocument $ parseMd text
+      updateFormState -- TODO: is the reason I had to put this here is that peek can't see events I send to the child component?
```

Comment:
Yeah, that's right - I didn't realise this was the case until you mentioned it now though. It is by design, but it's to prevent an infinite loop when queries to children are made from `peek` - `peek` is integrated into `eval` during `install`, so unfortunately the unobservability is also true for queries made from `eval`.